fastestimator/estimator.py

```diff
@@ -158,147 +191,141 @@ def _sort_traces(self) -> None:
-            AssertionError: If traces require inputs which are not available in the data dictionary.
+        Args:
+            mode: Current execution mode.
+
+        Returns:
+            List of schedulable items in estimator.
         """
-        for mode in self.pipeline.get_modes():
-            pipeline_all_outputs = self.pipeline.get_all_output_keys(mode, self.system.total_epochs)
-            network_all_outputs = self.network.get_all_output_keys(mode, self.system.total_epochs)
-            unmet_requirements = self.trace_inputs[mode] - (pipeline_all_outputs
-                                                            | network_all_outputs | self.trace_outputs[mode])
-            assert not unmet_requirements, "found missing key(s) during {}: {}".format(mode, unmet_requirements)
-
-    def _start_train(self) -> None:
+        return self.traces_in_use
+
+    def _start(self, run_modes: Set[str]) -> None:
         """The outer training loop.
 
         This method invokes the trace on_begin method, runs the necessary 'train' and 'eval' epochs, and then invokes
         the trace on_end method.
+
+        Args:
+            run_modes: The current execution modes.
         """
-        self._run_traces_on_begin({"train", "eval"})
-        try:
-            for self.system.epoch_idx in range(1, self.system.total_epochs + 1):
-                if "train" in self.pipeline.get_modes():
-                    self.system.mode = "train"
-                    self._run_epoch()
-                if "eval" in self.pipeline.get_modes():
-                    self.system.mode = "eval"
-                    self._run_epoch()
-        except EarlyStop:
-            pass  # On early stopping we still want to run the final traces and return results
-        self._run_traces_on_end({"train", "eval"})
-
-    def _start_test(self) -> None:
-        """The outer testing loop.
-
-        This method invokes the trace on_begin method, runs a 'test' epoch, and then invokes the trace on_end method.
-        """
-        self._run_traces_on_begin({"test"})
-        self._run_epoch()
-        self._run_traces_on_end({"test"})
+        all_traces = get_current_items(self.traces_in_use, run_modes=run_modes)
+        self._sort_traces(all_traces)
+        self._run_traces_on_begin(traces=all_traces)
+        if "train" in run_modes or "eval" in run_modes:
+            try:
+                for self.system.epoch_idx in range(1, self.system.total_epochs + 1):
+                    if "train" in self.pipeline.get_modes(epoch=self.system.epoch_idx):
+                        self.system.mode = "train"
+                        self._run_epoch()
+                    if "eval" in self.pipeline.get_modes(epoch=self.system.epoch_idx):
+                        self.system.mode = "eval"
+                        self._run_epoch()
+            except EarlyStop:
+                pass  # On early stopping we still want to run the final traces and return results
+        else:
```

Comment:
try/except needs to be outside of the top level if

Reply:
done